plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/parameterInfo/KotlinHighLevelFunctionParameterInfoHandler.kt

```diff
@@ -0,0 +1,451 @@
+        val GREEN_BACKGROUND: Color = JBColor(Color(231, 254, 234), Gray._100)
+
+        val STOP_SEARCH_CLASSES: Set<Class<out KtElement>> = setOf(
+            KtNamedFunction::class.java,
+            KtVariableDeclaration::class.java,
+            KtValueArgumentList::class.java,
+            KtLambdaArgument::class.java,
+            KtContainerNode::class.java,
+            KtTypeArgumentList::class.java
+        )
+    }
+
+    override fun getActualParameterDelimiterType(): KtSingleValueToken = KtTokens.COMMA
+
+    override fun getArgListStopSearchClasses(): Set<Class<out KtElement>> = STOP_SEARCH_CLASSES
+
+    override fun getArgumentListClass() = argumentListClass.java
+
+    override fun showParameterInfo(element: TArgumentList, context: CreateParameterInfoContext) {
+        context.showHint(element, element.textRange.startOffset, this)
+    }
+
+    override fun findElementForParameterInfo(context: CreateParameterInfoContext): TArgumentList? {
+        val file = context.file as? KtFile ?: return null
+
+        val token = file.findElementAt(context.offset) ?: return null
+        val argumentList = PsiTreeUtil.getParentOfType(token, argumentListClass.java, true, *STOP_SEARCH_CLASSES.toTypedArray())
+            ?: return null
+
+        val callElement = argumentList.parent as? KtElement ?: return null
+        return analyse(callElement) {
+            // TODO: FE 1.0 plugin collects all candidates (i.e., all overloads), even if arguments do not match. Not just resolved call.
+            // See Call.resolveCandidates() in core/src/org/jetbrains/kotlin/idea/core/Utils.kt. Note `replaceCollectAllCandidates(true)`.
+
+            val resolvedCall = when (callElement) {
+                is KtCallElement -> callElement.resolveCall()
+                is KtArrayAccessExpression -> callElement.resolveCall()
+                else -> return null
+            }
+
+            val candidates = resolvedCall?.targetFunction?.candidates ?: return null
+            context.itemsToShow = candidates.map { CandidateInfo(it.createPointer()) }.toTypedArray()
+
+            argumentList
+        }
+    }
+
+    override fun findElementForUpdatingParameterInfo(context: UpdateParameterInfoContext): TArgumentList? {
+        val element = context.file.findElementAt(context.offset) ?: return null
+        return PsiTreeUtil.getParentOfType(element, argumentListClass.java)
+    }
+
+    override fun updateParameterInfo(argumentList: TArgumentList, context: UpdateParameterInfoContext) {
+        if (context.parameterOwner !== argumentList) {
+            context.removeHint()
+        }
+        val parameterIndex = getParameterIndex(context, argumentList)
+        context.setCurrentParameter(parameterIndex)
+
+        val callElement = argumentList.parent as? KtElement ?: return
+        for (objectToView in context.objectsToView) {
+            val candidateInfo = objectToView as? CandidateInfo ?: continue
+            analyse(callElement) {
```

Comment:
Should this analyse block be hoisted out of the loop? Then I believe the first when branch can be moved out of the loop as well.

Reply:
Yes, that works. Thanks for the suggestion!